MusicGen Inference Endpoints custom handler tutorial (#1355)
huggingface/blog: musicgen-inference.md

---
title: "Deploy MusicGen using Inference Endpoints" 
thumbnail: /blog/assets/musicgen-inference/thumbnail.png
authors:
- user: reach-vb 
- user: merve
---

> <h1> Deploy MusicGen using Inference Endpoints </h1>

Reviewer: Should we name it something more catchy? `Run MusicGen with an API`? Or something along those lines?

Reviewer 2: @Vaibhavs10 I think the word "deploy" here is good for SEO, actually. To make it a bit catchier I would only slightly rephrase, maybe "Deploy MusicGen in no time with Inference Endpoints" (?)

<!-- {blog_metadata} -->
<!-- {authors} -->

[MusicGen](https://huggingface.co/docs/transformers/main/en/model_doc/musicgen) is a powerful music generation model that takes in text prompt and melody control and outputs music. In this blog post, we will walk you through how to deploy it using [Inference Endpoints](https://huggingface.co/inference-endpoints). 

`transformers` pipelines are powerful abstractions to do inference over `transformers` based models. The pipeline API also enables easy deployment of models in Inference Endpoints, with only few clicks. If the model we want to deploy in Inference Endpoints lacks pipeline, we can write our own inference function for it, called [custom handlers](https://huggingface.co/docs/inference-endpoints/guides/custom_handler). 

As the time this blog post is released, MusicGen lacks dedicated `transformers` pipeline. To implement customer handler for MusicGen and serve it, we will..
1. Duplicate the MusicGen repository we want to serve,
2. Write custom handler in `handler.py`, and the dependencies in `requirements.txt` and add them to the duplicated repository,
3. Create Inference Endpoint for that repository.

### Let's go!

> First, we will duplicate the [facebook/musicgen-small](https://huggingface.co/facebook/musicgen-small) repository to our own profile using [repository duplicator](https://huggingface.co/spaces/osanseviero/repo_duplicator).

Reviewer: Actually now they don't need to do anything, they can deploy with one-click: https://huggingface.co/reach-vb/musicgen-large-fp16-endpoint (and then clicking on deploy to inference endpoints)

```python
from transformers import AutoProcessor, MusicgenForConditionalGeneration
from datasets import load_dataset

processor = AutoProcessor.from_pretrained("facebook/musicgen-small")
model = MusicgenForConditionalGeneration.from_pretrained("facebook/musicgen-small")

dataset = load_dataset("sanchit-gandhi/gtzan", split="train", streaming=True)
sample = next(iter(dataset))["audio"]

# take the first half of the audio sample
sample["array"] = sample["array"][: len(sample["array"]) // 2]

inputs = processor(
    audio=sample["array"],
    sampling_rate=sample["sampling_rate"],
    text=["80s blues track with groovy saxophone"],
    padding=True,
    return_tensors="pt",
)
audio_values = model.generate(**inputs, do_sample=True, guidance_scale=3, max_new_tokens=256)
```

For `handler.py`, we can use [this template](https://huggingface.co/docs/inference-endpoints/guides/custom_handler#3-customize-endpointhandler) and override `__init__` and `__call__` methods to contain above code. `__init__` will initialize the model and the processor, and `__call__` will take the data and return the generated music. You can find the modified `EndpointHandler` class below. 👇 

> ```python

Reviewer: This needs to be updated with this: https://github.com/Vaibhavs10/deploy-audio-endpoints/blob/main/musicgen/handler.py - I made it work with `fp16` so it is a bit faster now. Also I am deploying a large model.

        # load model and processor from path
        self.processor = AutoProcessor.from_pretrained(path)
        self.model = MusicgenForConditionalGeneration.from_pretrained(path).to("cuda")

    def __call__(self, data: Dict[str, Any]) -> Dict[str, str]:
        """
        Args:
            data (:dict:):
                The payload with the text prompt and generation parameters.
        """
        # process input
        inputs = data.pop("inputs", data)
        parameters = data.pop("parameters", None)

        # preprocess
        inputs = self.processor(
            text=[inputs],
            padding=True,
            return_tensors="pt",).to("cuda")

        # pass inputs with all kwargs in data
        if parameters is not None:
            outputs = self.model.generate(**inputs, max_new_tokens=256, **parameters)
        else:
            outputs = self.model.generate(**inputs, max_new_tokens=256)

        # postprocess the prediction
        prediction = outputs[0].cpu().numpy()

        return [{"generated_text": prediction}]
```

Then, we will create a `requirements.txt` file that contains dependencies to be able to run above code. In this case, it is below.

> ````

Reviewer: Note on requirements, IE requires explicit version for installation. It has a pretty old version of transformers installed by default. https://github.com/Vaibhavs10/deploy-audio-endpoints/blob/main/musicgen/requirements.txt

![Create Endpoint](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/ie_musicgen/create_endpoint.png)

After creating the endpoint, it will be up and running. Then we can simply send a request to the endpoint.

![Endpoint Running](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/ie_musicgen/endpoint_running.png)

We can query the endpoint with below snippet.

```bash
curl URL_OF_ENDPOINT \
-X POST \
-d '{"inputs":"alt rock song"}' \
-H "Authorization: {YOUR_TOKEN_HERE}" \
-H "Content-Type: application/json"
```

We can see the following waveform sequence as output.
```
[{"generated_text":[[-0.024490159,-0.03154691,-0.0079551935,-0.003828604, ...]]}]
```

You can convert the generated sequence to audio however you want. In Python, you can use `scipy` to write it to a .wav file. 

> ```python

Reviewer: Lovely! This is fantastic! Is there a way for us to embed audio in the blog?

Reviewer 2: That would be really cool!

Reviewer 3: I think this should work: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio, same way we embed video in other posts: https://github.com/huggingface/blog/blob/main/falcon.md?plain=1#L77

---
title: "Deploy MusicGen using Inference Endpoints" 
thumbnail: /blog/assets/musicgen-inference/thumbnail.png
authors:
- user: reach-vb 
- user: merve
---

<h1> Deploy MusicGen using Inference Endpoints </h1>

<!-- {blog_metadata} -->
<!-- {authors} -->

> [MusicGen](https://huggingface.co/docs/transformers/main/en/model_doc/musicgen) is a powerful music generation model that takes in text prompt and melody control and outputs music. In this blog post, we will walk you through how to deploy it using [Inference Endpoints](https://huggingface.co/inference-endpoints). 

Reviewer 2: ```suggestion
[MusicGen](https://huggingface.co/docs/transformers/main/en/model_doc/musicgen) is a powerful music generation model that takes in text prompt and melody control and outputs music. In this blog post, we will guide you through the process of deploying it using [Inference Endpoints](https://huggingface.co/inference-endpoints). 
```

Reviewer 2: nit: "walk you through how to ... " doesn't sound right to me.

> `transformers` pipelines are powerful abstractions to do inference over `transformers` based models. The pipeline API also enables easy deployment of models in Inference Endpoints, with only few clicks. If the model we want to deploy in Inference Endpoints lacks pipeline, we can write our own inference function for it, called [custom handlers](https://huggingface.co/docs/inference-endpoints/guides/custom_handler). 

Reviewer 2: ```suggestion
`transformers` pipelines offer powerful abstractions to run inference with `transformers`-based models. The pipeline API also enables easy deployment of models in Inference Endpoints, with only few clicks. If the model we want to deploy in Inference Endpoints lacks a pipeline, we can write our own inference function for it, called [custom handler](https://huggingface.co/docs/inference-endpoints/guides/custom_handler). 
```

> As the time this blog post is released, MusicGen lacks dedicated `transformers` pipeline. To implement customer handler for MusicGen and serve it, we will..

Reviewer 2: ```suggestion
At the time of this blog post's release, MusicGen lacks a dedicated `transformers` pipeline. To implement a custom handler function for MusicGen and serve it, we will need to:
```

1. Duplicate the MusicGen repository we want to serve,
2. Write custom handler in `handler.py`, and the dependencies in `requirements.txt` and add them to the duplicated repository,
3. Create Inference Endpoint for that repository.

### Let's go!

First, we will duplicate the [facebook/musicgen-small](https://huggingface.co/facebook/musicgen-small) repository to our own profile using [repository duplicator](https://huggingface.co/spaces/osanseviero/repo_duplicator).

Then, from the duplicated repository, we will add `handler.py` and `requirements.txt`.
First, let's take a look at how to infer MusicGen.

You can initialize and use MusicGen with below code 👇 

```python
from transformers import AutoProcessor, MusicgenForConditionalGeneration

processor = AutoProcessor.from_pretrained("facebook/musicgen-small")
model = MusicgenForConditionalGeneration.from_pretrained("facebook/musicgen-small")

inputs = processor(
    text=["80s pop track with bassy drums and synth"],
    padding=True,
    return_tensors="pt",
)

> audio_values = model.generate(**inputs, do_sample=True, guidance_scale=3, max_new_tokens=256)

Reviewer 3: Could be challenging as this is not the purpose of this article. Maybe link to https://huggingface.co/blog/how-to-generate stating that the process follows the same principles as in text generation.

> audio_values = model.generate(**inputs, do_sample=True, guidance_scale=3, max_new_tokens=256)

Reviewer: Added 🤗

```

> You can also input a melody to control the output like below. 

Reviewer 2: I would add a sentence or two explaining what adding a melody does to the output.

Reviewer: Done!

```python
from transformers import AutoProcessor, MusicgenForConditionalGeneration
from datasets import load_dataset

processor = AutoProcessor.from_pretrained("facebook/musicgen-small")
model = MusicgenForConditionalGeneration.from_pretrained("facebook/musicgen-small")

dataset = load_dataset("sanchit-gandhi/gtzan", split="train", streaming=True)
sample = next(iter(dataset))["audio"]

# take the first half of the audio sample
sample["array"] = sample["array"][: len(sample["array"]) // 2]

inputs = processor(
    audio=sample["array"],
    sampling_rate=sample["sampling_rate"],
    text=["80s blues track with groovy saxophone"],
    padding=True,
    return_tensors="pt",
)
audio_values = model.generate(**inputs, do_sample=True, guidance_scale=3, max_new_tokens=256)
```

> For `handler.py`, we can use [this template](https://huggingface.co/docs/inference-endpoints/guides/custom_handler#3-customize-endpointhandler) and override `__init__` and `__call__` methods to contain above code. `__init__` will initialize the model and the processor, and `__call__` will take the data and return the generated music. You can find the modified `EndpointHandler` class below. 👇 

Reviewer 2: ```suggestion
For `handler.py`, we can use [this template](https://huggingface.co/docs/inference-endpoints/guides/custom_handler#3-customize-endpointhandler) and override `__init__` and `__call__` methods to contain the above code. `__init__` will initialize the model and the processor, and `__call__` will take the data and return the generated music. You can find the modified `EndpointHandler` class below. 👇 
```

        """
        Args:
            data (:dict:):
                The payload with the text prompt and generation parameters.
        """
        # process input
        inputs = data.pop("inputs", data)
        parameters = data.pop("parameters", None)

        # preprocess
        inputs = self.processor(
            text=[inputs],
            padding=True,
            return_tensors="pt",).to("cuda")

        # pass inputs with all kwargs in data
        if parameters is not None:
            outputs = self.model.generate(**inputs, max_new_tokens=256, **parameters)
        else:
            outputs = self.model.generate(**inputs, max_new_tokens=256)

        # postprocess the prediction
        prediction = outputs[0].cpu().numpy()

        return [{"generated_text": prediction}]
```

Then, we will create a `requirements.txt` file that contains dependencies to be able to run above code. In this case, it is below.

```
transformers
accelerate
```

Then, simply uploading these two files to our repository will suffice to serve the model.

![inference-files](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/ie_musicgen/files.png)

> We can now create the Inference Endpoint. Simply head to [Inference Endpoints](https://huggingface.co/inference-endpoints) page and click `Deploy your first model`. Enter the model repository to point at your duplicated repository identifier, select the hardware and create the endpoint. Any hardware with minimum of 16 GB RAM should work for `musicgen-small`.

Reviewer 2: nit: "simply" again. I think you can drop it in this sentence.

---
title: "Deploy MusicGen using Inference Endpoints" 
thumbnail: /blog/assets/musicgen-inference/thumbnail.png
authors:
- user: reach-vb 
- user: merve
---

> <h1> Deploy MusicGen in no time with Inference Endpoints </h1>

Reviewer 3: How about `Use Inference Endpoints custom handlers to deploy MusicGen in no time`? (Not sure, but I think the custom handler part is important as it's the main lesson here).

Reviewer: @MKhalusova suggested: `Deploy MusicGen in no time with Inference Endpoints` WDYT?

Reviewer: I think it sounds better and I'd like to remove the `custom handlers` bit from the title. It makes it sound way more complex hehe.

<!-- {blog_metadata} -->
<!-- {authors} -->

> [MusicGen](https://huggingface.co/docs/transformers/main/en/model_doc/musicgen) is a powerful music generation model that takes in text prompt and melody control and outputs music. This blog post will guide you through deploying it using [🤗 Inference Endpoints](https://huggingface.co/inference-endpoints). 

Reviewer 4: ```suggestion
[MusicGen](https://huggingface.co/docs/transformers/main/en/model_doc/musicgen) is a powerful music generation model that takes in text prompt and melody control and outputs music. This blog post will guide you through deploying it using [Inference Endpoints](https://huggingface.co/inference-endpoints). 
```

> `transformers` pipelines offer powerful abstractions to run inference with `transformers`-based models. The pipeline API enables easy deployment of models in Inference Endpoints with only a few clicks. Suppose the model we want to deploy in Inference Endpoints lacks a pipeline, or you wish to deploy a non-transformers model. We can write our inference function, called [custom handler](https://huggingface.co/docs/inference-endpoints/guides/custom_handler). 
> 
> At the time of this blog post's release, MusicGen lacks a dedicated `transformers` pipeline. To implement a custom handler function for MusicGen and serve it, we will need to:
> 1. Duplicate the MusicGen repository we want to serve,
> 2. Write a custom handler in `handler.py` and the dependencies in `requirements.txt` and add them to the duplicated repository,
> 3. Create Inference Endpoint for that repository.
> 4. Skip steps 1-3 and deploy via a [custom MusicGen model repo](https://huggingface.co/reach-vb/musicgen-large-fp16-endpoint).

Reviewer 4: E.g.

```suggestion
Inference Endpoints allows us to write custom inference functions called [custom handler](https://huggingface.co/docs/inference-endpoints/guides/custom_handler). These are particularly useful when a model is not supported out-of-the-box by the `transformers` high-level abstraction `pipeline`.

At the time of this blog post's release, MusicGen lacks a dedicated `transformers` pipeline. To implement a custom handler function for MusicGen and serve it, we will need to:

At the time of this blog post's release, MusicGen lacks a dedicated `transformers` pipeline. To implement a custom handler function for MusicGen and serve it, we will need to:

1. Duplicate the MusicGen repository we want to serve,
2. Write a custom handler in `handler.py` and the dependencies in `requirements.txt` and add them to the duplicated repository,
3. Create Inference Endpoint for that repository.

We've done this for you already, so you can skip steps 1-3 and deploy via a [custom MusicGen model repo](https://huggingface.co/reach-vb/musicgen-large-fp16-endpoint) we built.
```

> 4. Skip steps 1-3 and deploy via a [custom MusicGen model repo](https://huggingface.co/reach-vb/musicgen-large-fp16-endpoint).

Reviewer 3: ```suggestion

Or simply use the final result and deploy our [custom MusicGen model repo](https://huggingface.co/reach-vb/musicgen-large-fp16-endpoint), where we just followed the steps above :)
```

I found it slightly confusing, as the focus of the post seems to be the process, not the deployment itself.

> 4. Skip steps 1-3 and deploy via a [custom MusicGen model repo](https://huggingface.co/reach-vb/musicgen-large-fp16-endpoint).
> ### Let's go!

Reviewer 4: ```suggestion
4. Skip steps 1-3 and deploy via a [custom MusicGen model repo](https://huggingface.co/reach-vb/musicgen-large-fp16-endpoint).

### Let's go!
```

> ### Let's go!

Reviewer 4: I wonder if we can move 1-3 to a lower visibility of the blog post and focus on the inference endpoints side of things

Reviewer: Yeah! I think the focus is for people to understand custom handlers better.

> First, we will duplicate the [facebook/musicgen-small](https://huggingface.co/facebook/musicgen-small) repository to our own profile using [repository duplicator](https://huggingface.co/spaces/osanseviero/repo_duplicator).

Reviewer: @osanseviero - Is there a better way to duplicate than that space? :p

Reviewer 4: No, that is the easiest way

> Then, we will add `handler.py` and `requirements.txt` from the duplicated repository.

Reviewer 3: ```suggestion
Then, we will add `handler.py` and `requirements.txt` to the duplicated repository.
```

> First, let's take a look at how to infer MusicGen.

Reviewer 4: ```suggestion
First, let's take a look at how to run inference with MusicGen.
```

> First, let's take a look at how to infer MusicGen.

Reviewer 3: ```suggestion
First, let's take a look at how to run inference with MusicGen.
```

> 
> You can initialize and use MusicGen with the below code 👇 

Reviewer 3: ```suggestion
```

Might be a bit superfluous with the above.

model = MusicgenForConditionalGeneration.from_pretrained("facebook/musicgen-small")

inputs = processor(
    text=["80s pop track with bassy drums and synth"],
    padding=True,
    return_tensors="pt",
)
audio_values = model.generate(**inputs, do_sample=True, guidance_scale=3, max_new_tokens=256)
```

You can also input a melody to control the output like below. 

```python
from transformers import AutoProcessor, MusicgenForConditionalGeneration
from datasets import load_dataset

processor = AutoProcessor.from_pretrained("facebook/musicgen-small")
model = MusicgenForConditionalGeneration.from_pretrained("facebook/musicgen-small")

dataset = load_dataset("sanchit-gandhi/gtzan", split="train", streaming=True)
sample = next(iter(dataset))["audio"]

# take the first half of the audio sample
sample["array"] = sample["array"][: len(sample["array"]) // 2]

inputs = processor(
    audio=sample["array"],
    sampling_rate=sample["sampling_rate"],
    text=["80s blues track with groovy saxophone"],
    padding=True,
    return_tensors="pt",
)
audio_values = model.generate(**inputs, do_sample=True, guidance_scale=3, max_new_tokens=256)

> ```

Reviewer: [To-Do]: VB - Add an audio sample here.

> For `handler.py`, we can use [this template](https://huggingface.co/docs/inference-endpoints/guides/custom_handler#3-customize-endpointhandler) and override `__init__` and `__call__` methods to contain the above code. `__init__` will initialize the model and the processor, and `__call__` will take the data and return the generated music. You can find the modified `EndpointHandler` class below. 👇 

Reviewer 3: ```suggestion
For `handler.py`, we can use [this template](https://huggingface.co/docs/inference-endpoints/guides/custom_handler#3-customize-endpointhandler) and override the `__init__` and `__call__` methods with our custom inference code. `__init__` will initialize the model and the processor, and `__call__` will take the data and return the generated music. You can find the modified `EndpointHandler` class below. 👇 
```

class EndpointHandler:
    def __init__(self, path=""):
        # load model and processor from path
        self.processor = AutoProcessor.from_pretrained(path)
        self.model = MusicgenForConditionalGeneration.from_pretrained(path, torch_dtype=torch.float16).to("cuda")

    def __call__(self, data: Dict[str, Any]) -> Dict[str, str]:
        """
        Args:
            data (:dict:):
                The payload with the text prompt and generation parameters.
        """
        # process input
        inputs = data.pop("inputs", data)
        parameters = data.pop("parameters", None)

        # preprocess
        inputs = self.processor(
            text=[inputs],
            padding=True,
            return_tensors="pt",).to("cuda")

        # pass inputs with all kwargs in data
        if parameters is not None:
            with torch.autocast("cuda"):
                outputs = self.model.generate(**inputs, **parameters)
        else:
            with torch.autocast("cuda"):
                outputs = self.model.generate(**inputs,)

        # postprocess the prediction
        prediction = outputs[0].cpu().numpy().tolist()

        return [{"generated_audio": prediction}]

> ```

Reviewer 3: So we are not using the melody control feature, are we? Maybe clarify that the inference endpoints handler receives the inputs for a text-generation task (if that's the case), and we don't have additional data?

Author: I addressed your comment.

> Then, we will create a `requirements.txt` file containing dependencies to run the above code. In this case, it is below.

Reviewer 3: ```suggestion
Then, we will create a `requirements.txt` file containing the dependencies we need to run the code:
```

```
transformers==4.31.0
accelerate>=0.20.3
```

> Then, simply uploading these two files to our repository will suffice to serve the model.

Reviewer 3: ```suggestion
Uploading these two files to our repository will suffice to serve the model.
```

![inference-files](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/ie_musicgen/files.png)

> We can now create the Inference Endpoint. Head to [Inference Endpoints](https://huggingface.co/inference-endpoints) page and click `Deploy your first model`. Enter the model repository to point at your duplicated repository identifier, select the hardware and create the endpoint. Any hardware with a minimum of 16 GB RAM should work for `musicgen-small`.

Reviewer 4: This is the part i would suggest to highlight more by having it first and using the custom model repo created by us. The less friction/simpler we can make this blog post the better

Reviewer: Do the current changes look good to you? :)

> We can now create the Inference Endpoint. Head to [Inference Endpoints](https://huggingface.co/inference-endpoints) page and click `Deploy your first model`. Enter the model repository to point at your duplicated repository identifier, select the hardware and create the endpoint. Any hardware with a minimum of 16 GB RAM should work for `musicgen-small`.

Reviewer 3: ```suggestion
We can now create the Inference Endpoint. Head to [Inference Endpoints](https://huggingface.co/inference-endpoints) page and click `Deploy your first model`. In the "Model repository" field, enter the identifier of your duplicated repository. Then select the hardware you want and create the endpoint. Any instance with a minimum of 16 GB RAM should work for `musicgen-small`.
```

![Create Endpoint](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/ie_musicgen/create_endpoint.png)

> After creating the endpoint, it will be up and running. Then we can simply send a request to the endpoint.

Reviewer 3: ```suggestion
After creating the endpoint, it will be automatically launched and ready to receive requests.
```

![Endpoint Running](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/ie_musicgen/endpoint_running.png)

We can query the endpoint with the below snippet.

> ```bash

Reviewer: We should also add a similar snippet on how to do it with Python. Ping the API w/ Python.

Author: `InferenceClient` is acting up a bit for my endpoint, will see if there's a bug.

curl URL_OF_ENDPOINT \
-X POST \
-d '{"inputs":"alt rock song"}' \
-H "Authorization: {YOUR_TOKEN_HERE}" \
-H "Content-Type: application/json"
```

We can see the following waveform sequence as output.
```
[{"generated_audio":[[-0.024490159,-0.03154691,-0.0079551935,-0.003828604, ...]]}]
```

You can convert the generated sequence to audio however you want. You can use `scipy` in Python to write it to a .wav file. 

```python
import scipy
import numpy as np

output = [{"generated_audio":[[-0.024490159,-0.03154691,-0.0079551935,-0.003828604, ...]]}]

> scipy.io.wavfile.write("musicgen_out.wav", rate=sampling_rate, data=np.array(output[0]["generated_audio"][0]))

Reviewer 4: Is it possible to display some audios in the blog post?

Reviewer: Yes! working on that now.

```

And voila! 

## Conclusion

> In this blog post, we have shown you how to deploy MusicGen using Inference Endpoints using a custom handler. The custom handler can be used for MusicGen and any model on Hugging Face Hub with no pipeline you wish to deploy. All you have to do is override the `Endpoint Handler` class in `handler.py` and add `requirements.txt` to have your project's dependencies. 

Reviewer 4: Maybe add in some places "how to generate music with AI" or similar, as that's what people will search

Reviewer: Added it at the top.

> In this blog post, we have shown you how to deploy MusicGen using Inference Endpoints using a custom handler. The custom handler can be used for MusicGen and any model on Hugging Face Hub with no pipeline you wish to deploy. All you have to do is override the `Endpoint Handler` class in `handler.py` and add `requirements.txt` to have your project's dependencies. 

Reviewer 3: ```suggestion
In this blog post, we have shown how to deploy MusicGen using Inference Endpoints with a custom inference handler. The same technique can be used for any other model in the Hub that does not have an associated pipeline. All you have to do is override the `Endpoint Handler` class in `handler.py`, and add `requirements.txt` to reflect your project's dependencies. 
```

### Let's go!

First, we will duplicate the [facebook/musicgen-large](https://huggingface.co/facebook/musicgen-large) repository to our own profile using [repository duplicator](https://huggingface.co/spaces/huggingface-projects/repo_duplicator).

Then, we will add `handler.py` and `requirements.txt` to the duplicated repository.
First, let's take a look at how to run inference with MusicGen.

```python
from transformers import AutoProcessor, MusicgenForConditionalGeneration

processor = AutoProcessor.from_pretrained("facebook/musicgen-large")
model = MusicgenForConditionalGeneration.from_pretrained("facebook/musicgen-large")

inputs = processor(
    text=["80s pop track with bassy drums and synth"],
    padding=True,
    return_tensors="pt",
)
audio_values = model.generate(**inputs, do_sample=True, guidance_scale=3, max_new_tokens=256)
```

Let's hear what it sounds like:

<audio>
<source src="https://huggingface.co/datasets/huggingface/documentation-images/blob/main/blog/ie_musicgen/musicgen_out.wav" type="audio/wav"> 
Your browser does not support the audio element. 
</audio> 

> Optionally, you can also condition the output with a melody. 

Reviewer: Done!

    def __init__(self, path=""):
        # load model and processor from path
        self.processor = AutoProcessor.from_pretrained(path)
        self.model = MusicgenForConditionalGeneration.from_pretrained(path, torch_dtype=torch.float16).to("cuda")

    def __call__(self, data: Dict[str, Any]) -> Dict[str, str]:
        """
        Args:
            data (:dict:):
                The payload with the text prompt and generation parameters.
        """
        # process input
        inputs = data.pop("inputs", data)
        parameters = data.pop("parameters", None)

        # preprocess
        inputs = self.processor(
            text=[inputs],
            padding=True,
            return_tensors="pt",).to("cuda")

        # pass inputs with all kwargs in data
        if parameters is not None:
            with torch.autocast("cuda"):
                outputs = self.model.generate(**inputs, **parameters)
        else:
            with torch.autocast("cuda"):
                outputs = self.model.generate(**inputs,)

        # postprocess the prediction
        prediction = outputs[0].cpu().numpy().tolist()

        return [{"generated_audio": prediction}]
```

> To keep things simple in this example we are only generating audio from text, and not conditioning it with a melody.

Reviewer 3: ```suggestion
To keep things simple, in this example we are only generating audio from text, and not conditioning it with a melody.
```
